### ISSUE TITLE
Null pointer in the protein inference due to custom FASTA headers

### ISSUE THREAD

**Issue by tivdnbos** [compomics/peptide-shaker-2.0-issue-tracker#46]
Hi,

First of all: great to see that the new version of PeptideShaker is out! :)
I started some searches last week Thursday, but since then the progress is still on 94% and the memory consumption stays around 95 GB RAM. The progress bar on the left also doesn't show a percentage. In the previous version of PeptideShaker, the memory consumption went over 200 GB RAM and crashed afterwards.
I have to do a fraction analysis of 68 mgf files (in total 2 million spectra), against a concatenated target/decoy metaproteomics database of 20 million proteins.

Cheers,
Tim

![PS2](https://user-images.githubusercontent.com/13612447/70387227-b0cf6680-19a2-11ea-8894-f0908e7921d6.png)

**Comment by mvaudel** [comment]
Hi Tim,

Thanks for giving the new versions a spin on this data set. Just eyeballing htop and the progress dialog, it seems that the tool is busy sorting the proteins - this is a quite demanding step but given that IO and RAM do not seem to be limiting it should merely be a matter of better using the CPUs. I will look into it, but things are quite hectic these days so I will most likely not be able to solve this before the end of January.

Apologies for this delay,

Marc

**Comment by hbarsnes** [comment]
Hi Tim,

Just to be sure, I assume you have checked that there are no exceptions indicated in the PeptideShaker log file? In other words, that the problem is in fact the processing being slow and not that it has simply stopped due to an uncaught exception?

Best regards,
Harald

**Comment by tivdnbos** [comment]
Hi,

Thanks for looking into this. Indeed, it was very slow, but didn't crash. I was wondering since the status bar didn't show any progress (nor 0, nor any other number), if the process actually got started. Any how, I cancelled the process myself.

Best,
Tim

**Comment by mvaudel** [comment]
This process is now taking better advantage of multiple CPUs.
https://github.com/compomics/peptide-shaker/commit/27df26ba844347a2558ed77c5c726e97e06687ca
I am not sure whether it solves the issue but at least things should go faster. Would it be possible for you to share the database with us for testing? And maybe a couple of mgfs?

**Comment by hbarsnes** [comment]
Hi Tim,

An updated beta release is now available: http://genesis.ugent.be/maven2/eu/isas/peptideshaker/PeptideShaker/2.0.0-beta/PeptideShaker-2.0.0-beta.zip.

Please give it a go an let us know if it solves the issue or not.

Best regards,
Harald

**Comment by tivdnbos** [comment]
Hi Harald, Marc,

I just stopped the loading of the search results after approx. 7 days of not getting by "Selecting leading proteins, inferring peptide and protein inference status"...

Best regards,
Tim

**Comment by hbarsnes** [comment]
Hi Tim,

And there were no exceptions in the PeptideShaker log file?

Best regards,
Harald

**Comment by hanne-hox** [comment]
Hi Harald!
I am the one giving Tim hopeless data sets to analyse. I also got the same issue as Tim decribes above with the new beta version and another dataset: the progress is halted at 86% even though it was quite efficiently getting up to there. I will leave the process running until tomorrow, but I am not too hopeful, considering the experiences Tim had? Edit: After 24 hours, it is still at the same spot, with no indication it is doing anything. the Memory usage is about 5% and CPU at about 12% of the available, so the system shouldn't be a limitation here. 

![image](https://user-images.githubusercontent.com/61194017/74741976-128f3500-525e-11ea-88dc-2e50c958f087.png)

The dataset is 23 files, I only searched against X!Tandem, however, I also have a sequence database created from sequence data in house, but it is not hopelessly large: 11683 proteins (twice as many in the concatenated one with decoys). 

[log-report.txt](https://github.com/compomics/peptide-shaker-2.0-issue-tracker/files/4219573/log-report.txt)

**Comment by hbarsnes** [comment]
Hi Hanne,

> And there were no exceptions in the PeptideShaker log file?

I meant the PeptideShaker error log. You'll find it via the Welcome dialog > Settings & Help > Help > Bug Report. :)

Best regards,
Harald

**Comment by hanne-hox** [comment]
Here is the bug report. :-) 
Seems like it has some issues with creating the project. 
(I have tried running it several times with different number of files, and even just one sample, but encounter the same issue every time in the beta-version. In version 1.16.44 I am able to get out results from one single file, but not multiples, which is cancelled due to memort restriction).

[PeptideShaker 2.0.0-beta log.txt](https://github.com/compomics/peptide-shaker-2.0-issue-tracker/files/4225240/PeptideShaker.2.0.0-beta.log.txt)

**Comment by hbarsnes** [comment]
There seems to be a bug in the protein inference:

```
Caused by: java.lang.NullPointerException
	at eu.isas.peptideshaker.protein_inference.ProteinInference.compareMainProtein(ProteinInference.java:464)
	at eu.isas.peptideshaker.protein_inference.ProteinInference.inferPiStatus(ProteinInference.java:175)
	at eu.isas.peptideshaker.protein_inference.ProteinInference.lambda$inferPiStatus$1(ProteinInference.java:115)
```

Which seems to be due to a formatting issue with your FASTA file. If you have non-standard headers I would urge you to use our non-standard header formatting as detailed here: https://github.com/compomics/searchgui/wiki/DatabaseHelp#non-standard-fasta. This should hopefully fix the problem. You will have to research the data though.

I'm also including an updated beta version that avoid the given null pointer above: https://www.dropbox.com/s/hwicgpvlcg624gt/PeptideShaker-2.0.0.1-beta.zip?dl=0. This is only a temporary fix though and may result in other issues down the line. But should allow you to load the data.

**Comment by hanne-hox** [comment]
Finally got around to test the issue. Used the 2.0.0.0-beta version still, and modified the headings to include the >generic|accession no. as you suggested. Now I was able to progress beyond the "Selecting leading proteins..." point. 

At the QC/validation point I ran out of memory and the process was terminated. I then did discover that the PS have reverted back to using only 1GB memory and the 32-bit version (I got the red Low memory warning on the front). I think this must have caused at least some of the problems I have been having previously, but the warning didn't show before now. 

When I provided 64-bit Java and 64GB memory it managed to run (yay).
